File: trie/word_search_2/src/main.rs
/*
Given a 2-D grid of characters board and a list of strings words, return all words that are present in the grid.

For a word to be present it must be possible to form the word with a path in the board with horizontally or vertically neighboring cells. The same cell may not be used more than once in a word.
*/

use std::collections::HashSet;

use crate::prefix_tree::Node;

mod prefix_tree;

struct Solution {}

impl Solution {
    fn dfs(
        r: i32, 
        c: i32, 
        mut node: &mut Node, 
        mut cur_str: String, 
        rows: i32, 
        cols: i32, 
        seen: &mut HashSet<(i32,i32)>, 
        board: &Vec<Vec<char>>,
        res: &mut HashSet<String>,
    ) 
    {
        if (r < 0 || r >= rows) || (c < 0 || c >= cols) || seen.contains(&(r,c)) {
            return;
        }

        let next_char = board[r as usize][c as usize];
        if !node.children.contains_key(&next_char) {
            return
        }

        node = node.children.get_mut(&next_char).expect("Key should exist");
        cur_str += &next_char.to_string();
        if node.is_end_of_word && !res.contains(&cur_str) {
            res.insert(cur_str.clone());
        }
        seen.insert((r,c));

        Self::dfs(r+1,c,node,cur_str.clone(),rows,cols,seen,board,res);
        Self::dfs(r-1,c,node,cur_str.clone(),rows,cols,seen,board,res);
        Self::dfs(r,c+1,node,cur_str.clone(),rows,cols,seen,board,res);
        Self::dfs(r,c-1,node,cur_str.clone(),rows,cols,seen,board,res);

        seen.remove(&(r,c));
    }

    pub fn find_words(board: Vec<Vec<char>>, words: Vec<String>) -> Vec<String> {
        let mut root = Node::new();
        let mut res: HashSet<String> = HashSet::new();

        for word in words {
            root.add_word(word);
        }

        let rows = board.len() as i32;
        let cols = board[0].len() as i32;

        for r in 0..rows {
            for c in 0..cols {
                Self::dfs(r,c,&mut root,"".to_string(),rows,cols,&mut HashSet::new(), &board, &mut res);
            }
        }

        res.into_iter().collect()
    }
}

fn main() {
    let board: Vec<Vec<char>> = vec![
      vec!['a','b','c','d'],
      vec!['s','a','a','t'],
      vec!['a','c','k','e'],
      vec!['a','c','d','n'],
    ];

    let board: Vec<Vec<char>> = vec![
        vec!['b','a'],
        vec!['o','d'],
    ];
    let words = vec!["cat".to_string(), "back".to_string(), "backend".to_string()];
    let words = vec!["bad".into(), "ad".into(), "badger".into()];
    dbg!(Solution::find_words(board, words));
}

File: trie/word_search_2/main.py
"""
Given a 2-D grid of characters board and a list of strings words, return all words that are present in the grid.

For a word to be present it must be possible to form the word with a path in the board with horizontally or vertically neighboring cells. The same cell may not be used more than once in a word.


SOLUTION:
lesson here is to not keep trying the same thing over and over again. I initially attempted building the trie based on the board, but you build it based 
on the words. 

Then you can easily do your DFS starting from each c in board, and basically 
"""


class Node:
    def __init__(self) -> None:
        self.is_word = False
        self.children = {}

    def __repr__(self) -> str:
        return f"{self.children}"

    def add_word(self, word: str) -> None:
        cur = self
        for c in word:
            if c not in cur.children:
                cur.children[c] = Node()
            cur = cur.children[c]

        cur.is_word = True

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        res = set()

        root = Node()

        for word in words:
            root.add_word(word)


        seen = set()
        ROWS, COLS = len(board), len(board[0])
        def dfs(r: int, c: int, cur: Node, cur_str: str):
            if r not in range(ROWS) or c not in range(COLS) or board[r][c] not in cur.children or (r,c) in seen:
                return
            val = board[r][c]
            seen.add((r,c))
            tmp = cur
            cur = cur.children[val]
            cur_str += val
            if cur.is_word and cur_str not in res:
                res.add(cur_str)
            dirs = [
                    [r+1,c],
                    [r-1,c],
                    [r,c+1],
                    [r,c-1],
            ]

            for dir in dirs:
                dfs(dir[0], dir[1], cur, cur_str)

            seen.remove((r,c))

        for r in range(ROWS):
            for c in range(COLS):
                dfs(r,c,root,"")

        return list(res)

sol = Solution()


board = [
  ["a","b","c","d"],
  ["s","a","a","t"],
  ["a","c","k","e"],
  ["a","c","d","n"]
]

board = [["o","a","b","n"],["o","t","a","e"],["a","h","k","r"],["a","f","l","v"]]

words = ["bat","cat","back","backend","stack"]

words = ["oa","oaa"]

print(sol.findWords(board,words))
